core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -604,6 +671,77 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
+        implicit transid: TransactionId): Future[Boolean] = {
+        // traverse all actions and "inline" all actions that are sequence
+        // keep track of all sequences to detect recursion
+        info(this, s"Checking limits and recursion for actions in a sequence $sequence $components")
+        // first check that out of the box no more components than necessary
+        if (components.size > whiskConfig.actionSequenceLimit.toInt)
+            Future.failed(new TooManyActionsInSequence())
+        val resolvedSequence = WhiskAction.resolveAction(entityStore, sequence) // this assumes that entityStore is the same for actions and packages
+        val resolvedComponents = components map { c => WhiskAction.resolveAction(entityStore, c) }
+        val futureResComponents = Future.sequence(resolvedComponents)
+        futureResComponents flatMap { resComponents =>
+            // start inlining while keeping track of all sequences to check for recursion
+            resolvedSequence flatMap { sequence =>
+                inlineComponentsAndCountAtomicActions(0, resComponents.toList, List(sequence))
+            }
+        }
+    }
+
+    /*
+     * helper function that traverses the actions to inline and counts atomic actions while keeping track of sequences
+     * @param atomicActionsCnt the atomic actions seen so far
+     * @param actionsToInline the actions to inline
+     * @param sequences the sequences encountered during inlining
+     * @return a successful future in case the conditions were met or a failed fugure with either too many actions or recursion found
+     */
+    private def inlineComponentsAndCountAtomicActions(atomicActionsCnt: Int, actionsToInline: List[FullyQualifiedEntityName], sequences: List[FullyQualifiedEntityName])(
+        implicit transid: TransactionId): Future[Boolean] = {
+        if (atomicActionsCnt > whiskConfig.actionSequenceLimit.toInt)
+            Future.failed(new TooManyActionsInSequence())
+        else {
+            actionsToInline match {
+                case action :: restActions =>
+                    val docid = DocId(action.toString)
+                    WhiskAction.get(entityStore, docid) flatMap { act =>
+                        act.exec match {
+                            case seq: SequenceExec =>
+                                // this is actually a sequence, check already traversed sequences
+                                if (sequences.contains(FullyQualifiedEntityName(act.namespace, act.name))) {
+                                   Future.failed(new SequenceWithRecursion())
+                                } else {
+                                    // resolve the components first
+                                    // need to inline each of its components
+                                    val resolvedComponents = seq.fixedComponents.get map { c => WhiskAction.resolveAction(entityStore, c) }
```

Comment:
Is it guarenteed that this doesn't throw an exception? In general `Option.get` should never be used.